### PR TITLE
feat: add `useAsync` and improve `ssrRef`

### PR DESCRIPTION
**Pull request by mathe42**
*Features*
 - add `useAsync` helper to prefetch data and send in ssrContext to client
 - updates `serverPrefetch` helper
 - improves `ssrRef` handling

### CODE REVIEW

**Comment by mathe42** [comment]
When this PR is accepted we have to rewrite fetch. I didn't look into that :D

**Comment by mathe42** [comment]
Finished for review.

File: src/fetch.ts [hunk 1]
```diff
@@ -1,9 +1,8 @@
 import Vue from 'vue'
 import { getCurrentInstance, onBeforeMount } from '@vue/composition-api'
+import { onServerPrefetch } from './server-prefetch'
 
-import { onServerPrefetch } from './ssr-ref'
-
-import { ComponentInstance } from '@vue/composition-api/dist/component'
+import type { ComponentInstance } from '@vue/composition-api/dist/component'
```
Comment (mathe42) [inline src/fetch.ts:5]: Only import types there is only a d.ts file

File: src/async.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    if (p instanceof Promise) {
+      if (process.server) {
+        onServerPrefetch(async () => {
+          _ref.value = await p
+        })
+      } else {
+        // TODO: handle router and catch
+        // eslint-disable-next-line
+        p.then(res => (_ref.value = res))
```
Comment (mathe42) [inline src/async.ts:20]: While pending do not change the route.

File: src/server-prefetch.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    })
+  }
+}
+
+export function onServerPrefetch(cb: () => any) {
+  prefetchFunctions.push(cb)
+  setupOnServerPrefetch()
+}
+
+export function onServerPrefetchEnd(cb: () => any) {
```
Comment (mathe42) [inline src/server-prefetch.ts:25]: register function that need to be called after the user functions

File: test/fixture/pages/ssr-ref.vue [hunk 4]
```diff
@@ -39,10 +44,21 @@ export default defineComponent({
       prefetchValue.value = await fetcher('result', 500)
     })
 
+    const asyncValue = useAsync(() =>
```
Comment (mathe42) [inline test/fixture/pages/ssr-ref.vue:47]: useAsync returns a ref with the result of the promise or null while pendig. router should wait until Promise is resolved beforr entering route (WIP)

File: src/async.ts (hunk 2 above)
Comment (mathe42) [reply src/async.ts:20]: WIP

**Comment by mathe42** [comment]
**Note:** The old and my new implementation do not allow to use outside of setup! I'm working on that there are some problems with `watch`. https://github.com/vuejs/composition-api/pull/311 might fix them.

File: src/ssr-ref.ts [hunk 5]
```diff
@@ -1,55 +1,46 @@
...
-}
-
-export const ssrRef = <T>(value: T, key?: string) => {
-  const val = ref<T>(getValue(value))
-
-  if (!key)
+/**
+ * Creates a Ref wich is in sync with the client.
+ */
+export const ssrRef = <T>(value: T, key?: string): Ref<T> => {
```
[the remark this answers is not in the document]
Comment (danielroe) [reply src/ssr-ref.ts:18]:
```suggestion
export const ssrRef = <T>(value: T | (() => T), key?: string): Ref<T> => {
```

File: src/async.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { ssrRef } from './ssr-ref'
+import { onServerPrefetch } from './server-prefetch'
+import { useContext } from './context'
+
+export const useAsync = <T>(cb: () => T | Promise<T>) => {
+  const { route } = useContext()
```
[the remark this answers is not in the document]
Comment (mathe42) [reply src/async.ts:6]: Agreed. we can add a key argument with the babel plugin.

File: src/async.ts (hunk 2 above)
Comment (mathe42) [reply src/async.ts:20]: Is not WIP anymore :D

File: test/fixture/pages/ssr-ref.vue [hunk 7]
```diff
@@ -39,10 +44,21 @@ export default defineComponent({
...
+      fetcher(process.server ? 'server' : 'client', 500)
+    )
+
+    // Error handeling
+    // useAsync(() => {
```
[the remark this answers is not in the document]
Comment (mathe42) [reply test/fixture/pages/ssr-ref.vue:51]: I can remove that comment if you use that code you can see how useAsync handles errors

File: src/ssr-ref.ts [hunk 8]
```diff
@@ -2,10 +2,7 @@ import { ref, Ref } from '@vue/composition-api'
 import { onServerPrefetchEnd } from './server-prefetch'
 
 function getValue<T>(value: T | (() => T)): T {
-  if (value instanceof Function) return value()
-  return value
-}
-  if (typeof value === 'function') return value()
+  if (typeof value === 'function') return (value as () => T)()
```
Comment (danielroe) [inline src/ssr-ref.ts:5]:
```suggestion
  if (value instanceof Function) return value()
```

File: src/babel.ts [hunk 9]
```diff
@@ -22,7 +22,11 @@ export default function ssrRefPlugin({ loadOptions, getEnv, types: t }: Babel) {
         }
       : {}),
     CallExpression(path) {
-      if (!('name' in path.node.callee) || path.node.callee.name !== 'ssrRef')
+      if (
+        !('name' in path.node.callee) ||
+        (path.node.callee.name !== 'ssrRef' &&
+          path.node.callee.name !== 'useAsync')
```
Comment (danielroe) [inline src/babel.ts:28]:
```suggestion
        !('name' in path.node.callee) ||
        !['ssrRef', 'useAsync'].includes(path.node.callee.name)
```

File: src/ssr-ref.ts [hunk 10]
```diff
@@ -1,55 +1,53 @@
...
+function clone<T>(obj: T): T {
+  if (typeof obj === 'object') {
+    return JSON.parse(JSON.stringify(obj))
+  } else {
+    return obj
+  }
 }
 
+/**
+ * Creates a Ref wich is in sync with the client.
```
Comment (danielroe) [inline src/ssr-ref.ts:24]:
```suggestion
 * Creates a Ref that is in sync with the client.
```

File: src/ssr-ref.ts [hunk 11]
```diff
@@ -1,55 +1,53 @@
...
+    const initVal = clone(val)
+    const sref = ref(val) as Ref<T>
 
-export const onServerPrefetch = (callback: Function) => {
-  prefetch(async () => {
-    await callback()
-    injectRefs()
-  })
+    onServerPrefetchEnd(() =>
+      initVal !== sref.value ? (data[key] = sref.value) : null
```
Comment (danielroe) [inline src/ssr-ref.ts:48]: I think we might need to do a comparison of stringified arrays/objects in this case.

File: src/ssr-ref.ts [hunk 12]
```diff
@@ -1,55 +1,53 @@
...
+    )
```
Comment (danielroe) [inline src/ssr-ref.ts:49]: I think these can be merged:
```
    const val = getValue(value)
    const initVal = clone(val)
    const _ref = ref(val) as Ref<T>

    onServerPrefetchEnd(() => {
      const finalVal = clone(_ref.value)
      if (initVal !== finalVal) data[key] = _ref.value
    })

    return _ref
```

File: src/async.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
+import { ssrRef } from './ssr-ref'
+import { onServerPrefetch } from './server-prefetch'
+import { Ref, isRef } from '@vue/composition-api'
+
+export const useAsync = <T>(
+  cb: () => T | Promise<T>,
+  key?: string | Ref<null>
+): Ref<null | T> => {
+  const _ref = isRef(key) ? key : ssrRef<T | null>(null, key)
```
Comment (danielroe) [inline src/async.ts:9]:
```suggestion
  if (!key) {
    throw new Error(
      "You must provide a key. You can have it generated automatically by adding 'nuxt-composition-api/babel' to your Babel plugins."
    )
  }
  
  const _ref = isRef(key) ? key : ssrRef<T | null>(null, key)
```

File: src/ssr-ref.ts (hunk 11 above)
Comment (danielroe) [reply src/ssr-ref.ts:48]: see below.

File: package.json [hunk 14]
```diff
@@ -43,7 +43,7 @@
     "release": "release-it",
     "test": "run-s test:*",
     "test:types": "tsd",
-    "test:e2e": "testcafe firefox:headless test/e2e --app 'node test/start-fixture.js'"
+    "test:e2e": "testcafe firefox:headless test/e2e --app node test/start-fixture.js"
```
Comment (danielroe) [inline package.json:46]:
```suggestion
    "test:e2e": "testcafe firefox:headless test/e2e --app 'node test/start-fixture.js'"
```

Comment (mathe42) [reply package.json:46]: Windows doesn't like that

File: src/ssr-ref.ts (hunk 12 above)
Comment (mathe42) [reply src/ssr-ref.ts:49]:
```
  const val = getValue(value)
    const initVal = clone(val)
    const _ref = ref(val) as Ref<T>

    onServerPrefetchEnd(() => {
      if (initVal !== _ref.value) data[key] = _ref.value
    })

    return _ref

```

Thus should work.

Comment (danielroe) [reply src/ssr-ref.ts:49]: 👍

**Comment by mathe42** [comment]
I don't know where the test fails. Lokaly `node test/start-fixture.js` doesn't work anymore:

```
ERROR  Failed to compile with 1 errors                                                                         friendly-errors 09:22:20


 ERROR  in ./lib/index.es.js                                                                                    friendly-errors 09:22:20

Cannot statically analyse 'require(…, …)' in line 461                                                           friendly-errors 09:22:20
                                                                                                                friendly-errors 09:22:20
 @ ./test/fixture/.nuxt/composition-api/plugin.js 7:0-53
 @ ./test/fixture/.nuxt/index.js
 @ ./test/fixture/.nuxt/client.js
 @ multi eventsource-polyfill webpack-hot-middleware/client?reload=true&timeout=30000&ansiColors=&overlayStyles=&name=client&path=/__webpack_hmr/client ./test/fixture/.nuxt/client.js

```

File: src/babel.ts [hunk 15]
```diff
@@ -22,7 +22,10 @@ export default function ssrRefPlugin({ loadOptions, getEnv, types: t }: Babel) {
         }
       : {}),
     CallExpression(path) {
-      if (!('name' in path.node.callee) || path.node.callee.name !== 'ssrRef')
+      if (
+        !('name' in path.node.callee) ||
+        ['ssrRef', 'useAsync'].includes(path.node.callee.name)
```
Comment (danielroe) [inline src/babel.ts:27]:
```suggestion
        !['ssrRef', 'useAsync'].includes(path.node.callee.name)
```

**Comment by danielroe** [comment]
> I don't know where the test fails. Lokaly `node test/start-fixture.js` doesn't work anymore:
> 
> ```
> ERROR  Failed to compile with 1 errors                                                                         friendly-errors 09:22:20
> 
> 
>  ERROR  in ./lib/index.es.js                                                                                    friendly-errors 09:22:20
> 
> Cannot statically analyse 'require(…, …)' in line 461                                                           friendly-errors 09:22:20
>                                                                                                                 friendly-errors 09:22:20
>  @ ./test/fixture/.nuxt/composition-api/plugin.js 7:0-53
>  @ ./test/fixture/.nuxt/index.js
>  @ ./test/fixture/.nuxt/client.js
>  @ multi eventsource-polyfill webpack-hot-middleware/client?reload=true&timeout=30000&ansiColors=&overlayStyles=&name=client&path=/__webpack_hmr/client ./test/fixture/.nuxt/client.js
> ```

You're getting the error because the change to the Babel plugin is now injecting a key into every function call, including (in this case) `require`.

**Comment by mathe42** [comment]
Why are these test failing?

**Comment by danielroe** [comment]
> Why are these test failing?

The reason they are now failing is that the e2e tests now run against a built version of the fixture, where this error is occuring: `Node.appendChild: Cannot add children to a Comment`.

**Comment by mathe42** [comment]
this
```html
<template>
  <div>
    <div>test: {{test}}</div>
  </div>
</template>

<script>
import {
  defineComponent,
  ref,
  onServerPrefetch
} from '@vue/composition-api'

export default defineComponent({
  setup() {
    const test = ref('test')

    onServerPrefetch(() => {
      test.value = 'Hello World.'
    })

    return {
      test
    }
  },
})
</script>
```

works

```html
<template>
  <div>
    <div>test: {{test}}</div>
  </div>
</template>

<script>
import {
  defineComponent,
  ref,
  onServerPrefetch
} from 'nuxt-composition-api'

export default defineComponent({
  setup() {
    const test = ref('test')

    onServerPrefetch(() => {
      test.value = 'Hello World.'
    })

    return {
      test
    }
  },
})
</script>

```
doesn't work...

**Comment by mathe42** [comment]
@danielroe had to rewrite onServerPrefetch now it works.

File: src/ssr-ref.ts [hunk 16]
```diff
@@ -1,53 +1,55 @@
...
+  } else {
+    const val = getValue(value)
+    const initVal = clone(val)
+    const _ref = ref(val) as Ref<T>
+
+    onServerPrefetchEnd(() => {
+      if (initVal !== _ref.value) {
+        data[key] = _ref.value
+      }
+    })
```
Comment (danielroe) [inline src/ssr-ref.ts:51]:
```suggestion
  const val = getValue(value)
  const initVal = clone(val)
  const _ref = ref(val) as Ref<T>

  onServerPrefetchEnd(() => {
    if (initVal !== _ref.value) data[key] = _ref.value
  })

  return _ref
```

File: src/ssr-ref.ts [hunk 17]
```diff
@@ -1,53 +1,49 @@
...
-    await callback()
-    injectRefs()
+  onServerPrefetchEnd(() => {
+    if (initVal !== _ref.value) data[key] = _ref.value
   })
+
+  return _ref
+
+    return _ref
+  }
```
Comment (danielroe) [inline src/ssr-ref.ts:48]:
```suggestion
```

**Comment by mathe42** [comment]
That was wanted when useing a function I think it should be in `__NUXT__`

File: src/ssr-ref.ts [hunk 18]
```diff
@@ -1,53 +1,47 @@
...
+  const val = getValue(value)
+  const initVal = clone(val)
+  const _ref = ref(val) as Ref<T>
 
-export const onServerPrefetch = (callback: Function) => {
-  prefetch(async () => {
-    await callback()
-    injectRefs()
+  onServerPrefetchEnd(() => {
+    if (initVal !== _ref.value) data[key] = _ref.value
```
Comment (danielroe) [inline src/ssr-ref.ts:42]:
```suggestion
    if (value instanceof Function || initVal !== _ref.value)
```

File: src/ssr-ref.ts [hunk 19]
```diff
@@ -1,53 +1,47 @@
...
+  const val = getValue(value)
+  const initVal = clone(val)
+  const _ref = ref(val) as Ref<T>
 
-export const onServerPrefetch = (callback: Function) => {
-  prefetch(async () => {
-    await callback()
-    injectRefs()
+  onServerPrefetchEnd(() => {
+    if (value instanceof Function || initVal !== _ref.value)
```
Comment (danielroe) [inline src/ssr-ref.ts:42]:
```suggestion
    if (value instanceof Function || initVal !== _ref.value)
      data[key] = _ref.value
```

**Comment by mathe42** [comment]
Looks good for me.

**Pull request merged**